perception/map_based_prediction/src/path_generator.cpp

```diff
@@ -0,0 +1,346 @@
+    geometry_msgs::msg::Pose object_frame_pose;
+    geometry_msgs::msg::Pose world_frame_pose;
+    object_frame_pose.position.x = object_twist.linear.x * dt;
+    object_frame_pose.position.y = object_twist.linear.y * dt;
+    tf2::Quaternion quat;
+    quat.setRPY(0.0, 0.0, 0.0);
+    object_frame_pose.orientation = tf2::toMsg(quat);
+    tf2::Transform tf_object2future;
+    tf2::Transform tf_world2object;
+    tf2::Transform tf_world2future;
+
+    tf2::fromMsg(object_pose, tf_world2object);
+    tf2::fromMsg(object_frame_pose, tf_object2future);
+    tf_world2future = tf_world2object * tf_object2future;
+    tf2::toMsg(tf_world2future, world_frame_pose);
+
+    path.path.push_back(world_frame_pose);
+  }
+
+  return path;
+}
+
+PredictedPath PathGenerator::generatePolynomialPath(
+  const std_msgs::msg::Header & origin_header, const TrackedObject & object,
+  const PosePath & ref_path)
+{
+  // Get current Frenet Point
+  const double ref_path_len = tier4_autoware_utils::calcArcLength(ref_path);
+  const auto current_point = getFrenetPoint(origin_header, object, ref_path);
+
+  // Step1. Set Target Frenet Point
+  // Note that we do not set position s,
+  // since we don't know the target longitudinal position
+  const double T = time_horizon_;
+  FrenetPoint terminal_point;
+  terminal_point.header = current_point.header;
+  terminal_point.header.stamp =
+    rclcpp::Time(current_point.header.stamp) + rclcpp::Duration::from_seconds(T);
+  terminal_point.s_vel = current_point.s_vel;
+  terminal_point.s_acc = 0.0;
+  terminal_point.d = 0.0;
+  terminal_point.d_vel = 0.0;
+  terminal_point.d_acc = 0.0;
+
+  // Step2. Generate Predicted Path on a Frenet coordinate
+  const auto frenet_predicted_path =
+    generateFrenetPath(current_point, terminal_point, ref_path_len);
+
+  // Step3. Interpolate Reference Path for converting predicted path coordinate
+  const auto interpolated_ref_path = interpolateReferencePath(ref_path, frenet_predicted_path);
+
+  if (frenet_predicted_path.size() < 2 || interpolated_ref_path.size() < 2) {
+    const PredictedPath empty_path;
+    return empty_path;
+  }
+
+  // Step4. Convert predicted trajectory from Frenet to Cartesian coordinate
+  return convertPredictedPath(object, frenet_predicted_path, interpolated_ref_path);
+}
+
+FrenetPath PathGenerator::generateFrenetPath(
+  const FrenetPoint & current_point, const FrenetPoint & target_point, const double max_length)
+{
+  FrenetPath path;
+  const double duration = rclcpp::Time(target_point.header.stamp).seconds() -
+                          rclcpp::Time(current_point.header.stamp).seconds();
+  const Eigen::Vector3d lat_coeff = calcLatCoefficients(current_point, target_point, duration);
+  const Eigen::Vector2d lon_coeff = calcLonCoefficients(current_point, target_point, duration);
+
+  path.reserve(static_cast<size_t>(duration / sampling_time_interval_));
+  for (double t = 0.0; t <= duration; t += sampling_time_interval_) {
```

Comment:
Is it a bit dangerous that path.size (`static_cast<size_t>(duration / sampling_time_interval_)`) and for loop cycle num `for (double t = 0.0; t <= duration; t += sampling_time_interval_) {` may be different?

Additionaly, IMO it may be ok not to reserve path vector beforehand?

Reply:
It just reserve the size of the vector, not 'resize' the size of the vector. So if won't be dangerous.

Reply:
Ok, then it's just for efficiency to reserve the capacity?
I'm not familiar with it, but do you think we should reserve the vector capacity before pushing back variables to the vector? ( I just wanted to know since not so many codes in Autoware use reserve before pushing back)

Reply:
Yes, it just literally reserves the capacity for efficiency. You can find more detailed explanations [here](https://suzulang.com/c-stdvector-resize%E3%81%A8reserve%E3%81%AE%E9%81%95%E3%81%84-%E5%8A%A0%E7%AD%86%E4%B8%AD%EF%BC%9F/). 

Push_back reserves the capacity less efficiently if we do not reserve the appropriate capacity in advance. I think other nodes also should use reserve function before executing push back function.

Reply:
Got it. Thanks!